server/src/main/java/org/elasticsearch/index/mapper/JsonFieldParser.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.apache.lucene.document.Field;
+import org.apache.lucene.index.IndexableField;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentParserUtils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A helper class for {@link JsonFieldMapper} parses a JSON object
+ * and produces an indexable field for each leaf value.
+ */
+public class JsonFieldParser {
+    private final MappedFieldType fieldType;
+    private final int ignoreAbove;
+
+    JsonFieldParser(MappedFieldType fieldType,
+                    int ignoreAbove) {
+        this.fieldType = fieldType;
+        this.ignoreAbove = ignoreAbove;
+    }
+
+    public List<IndexableField> parse(XContentParser parser) throws IOException {
+        XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT,
+            parser.currentToken(),
+            parser::getTokenLocation);
+
+        List<IndexableField> fields = new ArrayList<>();
+        int openObjects = 1;
+
+        while (true) {
+            if (openObjects == 0) {
```

Comment:
What happens if we're given malformed JSON?  If it's missing a closing brace, will this just loop forever?

Reply:
I had been wondering the same thing and have a test case for this: https://github.com/elastic/elasticsearch/pull/33923/files#diff-c8a7654deccf33995eeafb732768f18bR233

`XContentParser` will throw if it encounters an end token (or another non-sensical token) when it is expecting a closing brace.